module/zfs/dnode.c

```diff
@@ -1518,16 +1518,6 @@ dnode_hold_impl(objset_t *os, uint64_t object, int flag, int slots,
 		return (SET_ERROR(EINVAL));
 	}
 
-	if (dn->dn_free_txg) {
-		DNODE_STAT_BUMP(dnode_hold_free_txg);
-		type = dn->dn_type;
-		mutex_exit(&dn->dn_mtx);
-		dnode_slots_rele(dnc, idx, slots);
-		dbuf_rele(db, FTAG);
-		return (SET_ERROR((flag & DNODE_MUST_BE_ALLOCATED) ?
-		    ENOENT : EEXIST));
-	}
```

Comment:
That's interesting, you're right it does look like this is unreachable due to the early returns in both the `DNODE_MUST_BE_ALLOCATED` and `DNODE_MUST_BE_FREE` cases.  Then instead of simply removing this check let's replace it with an `ASSERT0(dn->dn_free_txg)` to make it clear these cases and handled above.  The dnode_hold_free_txg` entry should also be removed from the `dnode_stats` kstat.